File: go/pactman/sources.go
package pactman

import "reflect"

// The grouped views below are projections, not derivations: every field is
// copied 1:1 from one the API returned. Nothing here computes an "approved",
// "eligible" or "safe" verdict, and nothing infers a value from another field.
//
// Each accessor returns nil only when the API returned none of that source's
// fields at all. That keeps "the source was not returned" distinguishable from
// an explicit negative such as pub78_verified: false or a null field. Every
// accessor is safe to call on a nil *Nonprofit.

// Pub78Source holds the IRS Publication 78 findings.
type Pub78Source struct {
	Verified          *bool
	OrganizationName  *string
	EIN               *string
	City              *string
	State             *string
	Indicator         *string
	ChurchMessage     *string
	OrganizationTypes []*OrganizationType
	MostRecent        *string
}

// BMFSource holds the IRS Business Master File findings.
type BMFSource struct {
	Status                    *bool
	OrganizationName          *string
	EIN                       *string
	ChurchMessage             *string
	Subsection                *string
	SubsectionDescription     *string
	FoundationCode            *string
	FoundationCodeDescription *string
	FoundationTypeCode        *string
	FoundationTypeDescription *string
	Foundation509aStatus      *string
	RulingMonth               *string
	RulingYear                *string
	GroupExemption            *string
	ExemptStatusCode          *string
	FilingReqCode             *string
	MostRecent                *string
}

// AROESource holds the IRS Automatic Revocation of Exemption findings.
type AROESource struct {
	RevocationCode    *string
	RevocationDate    *string
	ReinstatementDate *string
}

// OFACSource holds the OFAC Specially Designated Nationals findings.
type OFACSource struct {
	// Status is the finding as the API phrases it. It is prose, not a flag;
	// the API does not return a boolean match indicator, and this SDK does not
	// invent one by matching on the wording.
	Status *string
}

// Wire fields behind each source, in the order the API reference lists them.
var (
	pub78Wire = []string{
		"pub78_verified", "pub78_organization_name", "pub78_ein", "pub78_city", "pub78_state",
		"pub78_indicator", "pub78_church_message", "organization_types", "most_recent_pub78",
	}
	bmfWire = []string{
		"bmf_status", "bmf_organization_name", "bmf_ein", "bmf_church_message", "bmf_subsection",
		"subsection_description", "foundation_code", "foundation_code_description",
		"foundation_type_code", "foundation_type_description", "foundation_509a_status",
		"ruling_month", "ruling_year", "group_exemption", "exempt_status_code", "filing_req_code",
		"most_recent_bmf",
	}
	aroeWire = []string{"revocation_code", "revocation_date", "reinstatement_date"}
	ofacWire = []string{"ofac_status"}
)

// Pub78 returns the Publication 78 findings, or nil if the API returned none.
func (n *Nonprofit) Pub78() *Pub78Source {
	if !n.returnedAny(pub78Wire) {
		return nil
	}

	return &Pub78Source{
		Verified:          n.Pub78Verified,
		OrganizationName:  n.Pub78OrganizationName,
		EIN:               n.Pub78EIN,
		City:              n.Pub78City,
		State:             n.Pub78State,
		Indicator:         n.Pub78Indicator,
		ChurchMessage:     n.Pub78ChurchMessage,
		OrganizationTypes: n.OrganizationTypes,
		MostRecent:        n.MostRecentPub78,
	}
}

// BMF returns the Business Master File findings, or nil if the API returned none.
func (n *Nonprofit) BMF() *BMFSource {
	if !n.returnedAny(bmfWire) {
		return nil
	}

	return &BMFSource{
		Status:                    n.BMFStatus,
		OrganizationName:          n.BMFOrganizationName,
		EIN:                       n.BMFEIN,
		ChurchMessage:             n.BMFChurchMessage,
		Subsection:                n.BMFSubsection,
		SubsectionDescription:     n.SubsectionDescription,
		FoundationCode:            n.FoundationCode,
		FoundationCodeDescription: n.FoundationCodeDescription,
		FoundationTypeCode:        n.FoundationTypeCode,
		FoundationTypeDescription: n.FoundationTypeDescription,
		Foundation509aStatus:      n.Foundation509aStatus,
		RulingMonth:               n.RulingMonth,
		RulingYear:                n.RulingYear,
		GroupExemption:            n.GroupExemption,
		ExemptStatusCode:          n.ExemptStatusCode,
		FilingReqCode:             n.FilingReqCode,
		MostRecent:                n.MostRecentBMF,
	}
}

// AROE returns the Automatic Revocation of Exemption findings, or nil if the
// API returned none.
func (n *Nonprofit) AROE() *AROESource {
	if !n.returnedAny(aroeWire) {
		return nil
	}

	return &AROESource{
		RevocationCode:    n.RevocationCode,
		RevocationDate:    n.RevocationDate,
		ReinstatementDate: n.ReinstatementDate,
	}
}

// OFAC returns the OFAC findings, or nil if the API returned none.
func (n *Nonprofit) OFAC() *OFACSource {
	if !n.returnedAny(ofacWire) {
		return nil
	}

	return &OFACSource{Status: n.OFACStatus}
}

// nonprofitFieldIndex maps a wire name to its field on Nonprofit.
var nonprofitFieldIndex = func() map[string]int {
	typ := reflect.TypeOf(Nonprofit{})
	index := make(map[string]int, typ.NumField())

	for i := 0; i < typ.NumField(); i++ {
		if name := jsonName(typ.Field(i)); name != "" {
			index[name] = i
		}
	}

	return index
}()

// returnedAny reports whether any of the wire fields is on the record: sent by
// the API (null included), or set on a record built in code.
func (n *Nonprofit) returnedAny(wire []string) bool {
	if n == nil {
		return false
	}

	value := reflect.ValueOf(n).Elem()

	for _, name := range wire {
		if n.Fields.Has(name) {
			return true
		}

		if i, ok := nonprofitFieldIndex[name]; ok && !value.Field(i).IsNil() {
			return true
		}
	}

	return false
}

File: ruby/lib/pactman/nonprofit_check_plus/sources.rb
# frozen_string_literal: true

module Pactman
  module NonprofitCheckPlus
    # IRS Publication 78 findings.
    class Pub78Source < Model
      field :verified, :boolean, null: true
      field :organization_name, String, null: true
      field :ein, String, null: true
      field :city, String, null: true
      field :state, String, null: true
      field :indicator, String, null: true
      field :church_message, String, null: true
      field :organization_types, [OrganizationType], null: true
      field :most_recent, String, null: true
    end

    # IRS Business Master File findings.
    class BmfSource < Model
      field :status, :boolean, null: true
      field :organization_name, String, null: true
      field :ein, String, null: true
      field :church_message, String, null: true
      field :subsection, String, null: true
      field :subsection_description, String, null: true
      field :foundation_code, String, null: true
      field :foundation_code_description, String, null: true
      field :foundation_type_code, String, null: true
      field :foundation_type_description, String, null: true
      field :foundation_509a_status, String, null: true
      field :ruling_month, String, null: true
      field :ruling_year, String, null: true
      field :group_exemption, String, null: true
      field :exempt_status_code, String, null: true
      field :filing_req_code, String, null: true
      field :most_recent, String, null: true
    end

    # IRS Automatic Revocation of Exemption findings.
    class AroeSource < Model
      field :revocation_code, String, null: true
      field :revocation_date, String, null: true
      field :reinstatement_date, String, null: true
    end

    # OFAC Specially Designated Nationals findings.
    class OfacSource < Model
      # The finding as the API phrases it. This is prose, not a flag; the API
      # does not currently return a boolean match indicator, and this gem does
      # not invent one by matching on the wording.
      field :status, String, null: true
    end

    # Grouped views over the source-specific findings on a {Nonprofit}.
    #
    # These are projections, not derivations: every attribute is copied 1:1 from
    # a field the API returned. Nothing here computes an "approved", "eligible"
    # or "safe" verdict, and nothing infers a value from another field.
    #
    # Each accessor returns `nil` when the API returned no data at all for that
    # source. That keeps "the source was not returned" distinguishable from an
    # explicit negative such as `pub78_verified: false` or a `null` field.
    module Sources
      PUB78 = {
        "verified" => "pub78_verified",
        "organization_name" => "pub78_organization_name",
        "ein" => "pub78_ein",
        "city" => "pub78_city",
        "state" => "pub78_state",
        "indicator" => "pub78_indicator",
        "church_message" => "pub78_church_message",
        "organization_types" => "organization_types",
        "most_recent" => "most_recent_pub78"
      }.freeze

      BMF = {
        "status" => "bmf_status",
        "organization_name" => "bmf_organization_name",
        "ein" => "bmf_ein",
        "church_message" => "bmf_church_message",
        "subsection" => "bmf_subsection",
        "subsection_description" => "subsection_description",
        "foundation_code" => "foundation_code",
        "foundation_code_description" => "foundation_code_description",
        "foundation_type_code" => "foundation_type_code",
        "foundation_type_description" => "foundation_type_description",
        "foundation_509a_status" => "foundation_509a_status",
        "ruling_month" => "ruling_month",
        "ruling_year" => "ruling_year",
        "group_exemption" => "group_exemption",
        "exempt_status_code" => "exempt_status_code",
        "filing_req_code" => "filing_req_code",
        "most_recent" => "most_recent_bmf"
      }.freeze

      AROE = {
        "revocation_code" => "revocation_code",
        "revocation_date" => "revocation_date",
        "reinstatement_date" => "reinstatement_date"
      }.freeze

      OFAC = {
        "status" => "ofac_status"
      }.freeze

      private_constant :PUB78, :BMF, :AROE, :OFAC

      class << self
        # Publication 78 findings, or `nil` if the API returned none.
        #
        # @param nonprofit [Nonprofit, Hash]
        # @return [Pub78Source, nil]
        def pub78(nonprofit)
          project(Pub78Source, nonprofit, PUB78)
        end

        # Business Master File findings, or `nil` if the API returned none.
        #
        # @param nonprofit [Nonprofit, Hash]
        # @return [BmfSource, nil]
        def bmf(nonprofit)
          project(BmfSource, nonprofit, BMF)
        end

        # Automatic Revocation of Exemption findings, or `nil` if the API returned none.
        #
        # @param nonprofit [Nonprofit, Hash]
        # @return [AroeSource, nil]
        def aroe(nonprofit)
          project(AroeSource, nonprofit, AROE)
        end

        # OFAC findings, or `nil` if the API returned none.
        #
        # @param nonprofit [Nonprofit, Hash]
        # @return [OfacSource, nil]
        def ofac(nonprofit)
          project(OfacSource, nonprofit, OFAC)
        end

        private

        # Copies the mapped fields onto a new object, preserving `nil` and `false`,
        # and the values themselves rather than copies of them.
        #
        # Returns `nil` only when every mapped field is absent from the response,
        # which is how "this source was not returned" is represented.
        def project(model, nonprofit, mapping)
          wire = wire_hash(nonprofit)
          projected = {}

          mapping.each do |target, wire_field|
            projected[target] = wire[wire_field] if wire.key?(wire_field)
          end

          projected.empty? ? nil : model.new(projected)
        end

        def wire_hash(nonprofit)
          case nonprofit
          when Nonprofit then nonprofit.to_h
          when Hash then nonprofit.transform_keys(&:to_s)
          else
            raise ArgumentError, "Expected a Nonprofit or a Hash, received #{nonprofit.class}."
          end
        end
      end
    end
  end
end
